File: src/Microsoft.AspNetCore.SignalR.Client/HttpConnectionOptions.cs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.

using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Net.WebSockets;
using System.Security.Cryptography.X509Certificates;
using Microsoft.AspNetCore.Http.Connections;

namespace Microsoft.AspNetCore.SignalR.Client
{
    public class HttpConnectionOptions
    {
        internal X509CertificateCollection _clientCertificates;
        internal IDictionary<string, string> _headers;
        internal CookieContainer _cookies;

        public Uri Url { get; set; }
        public TransportType? Transport { get; set; }
        public Func<HttpMessageHandler, HttpMessageHandler> MessageHandlerFactory { get; set; }
        public bool? UseDefaultCredentials { get; set; }
        public ICredentials Credentials { get; set; }
        public IWebProxy Proxy { get; set; }
        public Func<string> AccessTokenFactory { get; set; }
        public Action<ClientWebSocketOptions> WebSocketOptions { get; set; }

        public void AddClientCertificate(X509Certificate clientCertificate)
        {
            if (_clientCertificates == null)
            {
                _clientCertificates = new X509CertificateCollection();
            }

            _clientCertificates.Add(clientCertificate);
        }

        public void AddCookie(Cookie cookie)
        {
            if (_cookies == null)
            {
                _cookies = new CookieContainer();
            }

            _cookies.Add(cookie);
        }

        public void AddHeader(string name, string value)
        {
            if (_headers == null)
            {
                _headers = new Dictionary<string, string>();
            }

            _headers.Add(name, value);
        }
    }
}

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HttpConnectionOptions.java
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.

package com.microsoft.aspnet.signalr;

import java.util.function.Supplier;

public class HttpConnectionOptions {
    private Transport transport;
    private LogLevel loglevel;
    private Logger logger;
    private boolean skipNegotiate;
    private Supplier<String> accessTokenFactory;
    private HttpClient client;

    public HttpConnectionOptions() {}

    public HttpConnectionOptions(Transport transport, LogLevel logLevel, boolean skipNegotiate) {
        this.transport = transport;
        this.skipNegotiate = skipNegotiate;
        this.loglevel = logLevel;
    }

    public HttpConnectionOptions(Transport transport, Logger logger, boolean skipNegotiate) {
        this.transport = transport;
        this.skipNegotiate = skipNegotiate;
        this.logger = logger;
    }

    public void setTransport(Transport transport) {
        this.transport = transport;
    }

    public void setLoglevel(LogLevel loglevel) {
        this.loglevel = loglevel;
    }

    public void setSkipNegotiate(boolean skipNegotiate) {
        this.skipNegotiate = skipNegotiate;
    }

    public Transport getTransport() {
        return transport;
    }

    public LogLevel getLoglevel() {
        return loglevel;
    }

    public boolean getSkipNegotiate() {
        return skipNegotiate;
    }

    public Logger getLogger() {
        return logger;
    }

    public void setLogger(Logger logger) {
        this.logger = logger;
    }

    public void setAccessTokenFactory(Supplier<String> accessTokenFactory) {
        this.accessTokenFactory = accessTokenFactory;
    }

    public Supplier<String> getAccessTokenFactory() {
        return accessTokenFactory;
    }

    void setHttpClient(HttpClient client) {
        this.client = client;
    }

    HttpClient getHttpClient() {
        return client;
    }
}
